Enable temperature sensor on rt700

ZhaoxiangJin opened the pull request:
1. Enable nxp pmc temperature sensor (pmc-tmpsns) driver.
2. enable pmc-tmpsns on RT700
3. tested with samples/sensor/die_temp_polling
test result:
```
*** Booting Zephyr OS build v4.2.0-3691-gd3f239dcd8d6 ***
CPU Die temperature[temp0]: 25.8 °C
CPU Die temperature[temp0]: 25.6 °C
CPU Die temperature[temp0]: 24.9 °C
CPU Die temperature[temp0]: 25.2 °C
CPU Die temperature[temp0]: 25.3 °C
CPU Die temperature[temp0]: 25.6 °C
CPU Die temperature[temp0]: 25.1 °C
CPU Die temperature[temp0]: 25.1 °C
CPU Die temperature[temp0]: 25.1 °C
CPU Die temperature[temp0]: 25.1 °C
```

ZhaoxiangJin committed:
modules: hal_nxp: Pull in romapi for RT7xx

Pull in romapi for RT7xx

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: sensor: Enable NXP pmc tmpsns driver

This commit introduced NXP pmc temperature sensor
(pmc-tmpsns) driver.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
rebased to fix conflict

ZhaoxiangJin commented:
rebased to fix conflict

decsny reviewed (commented):
looks fine just one question

ZhaoxiangJin commented on drivers/sensor/nxp/nxp_pmc_tmpsns/nxp_pmc_tmpsns.c:
```diff
@@ -0,0 +1,156 @@
+		return ret;
+	}
+
+	ret = otp_fuse_read(CONFIG_NXP_PMC_TMPSNS_CALIBRATION_OTP_FUSE_INDEX,
```
This function is defined in [fsl_romapi_opt.h](https://github.com/zephyrproject-rtos/hal_nxp/blob/ecaecded549eb80914a62d3f947681e5680dd48e/mcux/mcux-sdk-ng/devices/RT/RT700/MIMXRT798S/drivers/romapi/otp/fsl_romapi_otp.c#L55). This PR requires romapi_opt update to the latest, so there is a [hal_nxp PR](https://github.com/zephyrproject-rtos/hal_nxp/pull/601)

ZhaoxiangJin commented on drivers/sensor/nxp/nxp_pmc_tmpsns/nxp_pmc_tmpsns.c:
```diff
@@ -0,0 +1,156 @@
+	static struct nxp_pmc_tmpsns_data _CONCAT(data, inst) = {			\
+		.buffer = 0,								\
+		.pmc_tmpsns_calibration = 0,						\
+		.pmc_tmpsns_value = 0,							\
```
updated

ZhaoxiangJin commented:
Update to fix conflicts

ZhaoxiangJin commented:
Hello @MaureenHelm, could you please revisit this pr? I think all your comments have been fixed. Thanks!

MaureenHelm commented on dts/arm/nxp/nxp_rt7xx_cm33_cpu0.dtsi:
```diff
@@ -1105,6 +1109,31 @@
 		status = "disabled";
 	};
+
+	mbox4: mailbox@189000 {
+		#mbox-cells = <1>;
+
+		compatible = "nxp,mbox-imx-mu";
+		reg = <0x189000 0x1000>;
+
+		interrupts = <31 0>;
+		rx-channels = <4>;
+
+		status = "disabled";
+	};
+
+	pmc_tmpsns: pmc-tmpsns {
+		compatible = "nxp,pmc-tmpsns";
+		status = "disabled";
+	};
+};
+
+&systick {
+	/*
+	 * RT700 cm33_cpu0 relies by default on the OS Timer for system
+	 * clock implementation, so the SysTick node is not to be enabled.
+	 */
+	status = "disabled";
```
Some extra stuff is being added that looks unrelated to the sensor. Rebasing mistake?

ZhaoxiangJin replied:
sorry for that, it would be a accident, i will update shortly

ZhaoxiangJin committed:
dts: nxp: add pmc-tmpsns to rt7xx dts

1. add pmc-tmpsns to rt7xx dts
2. add pmc-tmpsns to rt700 evk

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

jeppenodgaard commented on drivers/sensor/nxp/nxp_pmc_tmpsns/nxp_pmc_tmpsns.c:
```diff
@@ -0,0 +1,140 @@
+		return -ENOTSUP;
+	}
+
+	for (uint8_t index = 0; index < 15; ++index) {
```
```suggestion
	for (uint8_t index = 0; index < sizeof(pmc_tmpsns_select); ++index) {
```

ZhaoxiangJin replied:
updated

jeppenodgaard commented on drivers/sensor/nxp/nxp_pmc_tmpsns/nxp_pmc_tmpsns.c:
```diff
@@ -0,0 +1,140 @@
+		return -ENOTSUP;
+	}
+
+	return sensor_value_from_double(val, data->pmc_tmpsns_value);
```
```suggestion
	return sensor_value_from_float(val, data->pmc_tmpsns_value);
```

ZhaoxiangJin replied:
updated

jeppenodgaard commented on drivers/sensor/nxp/nxp_pmc_tmpsns/nxp_pmc_tmpsns.c:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	ret = otp_fuse_read(CONFIG_NXP_PMC_TMPSNS_CALIBRATION_OTP_FUSE_INDEX,
+			&(data->pmc_tmpsns_calibration));
```
```suggestion
			&data->pmc_tmpsns_calibration);
```

ZhaoxiangJin replied:
updated

jeppenodgaard commented on drivers/sensor/nxp/nxp_pmc_tmpsns/nxp_pmc_tmpsns.c:
```diff
@@ -0,0 +1,140 @@
+};
+
+#define NXP_PMC_TMPSNS_INIT(inst)							\
+	static struct nxp_pmc_tmpsns_data _CONCAT(data, inst);				\
```
Normally `config` and `data` static names are prepended with driver name.
Not sure if it is required. I think it might be annoying if they aren't during debugging.

ZhaoxiangJin replied:
just curious, what are the benefits of having the driver name prefixed during debugging?

jeppenodgaard replied:
For watchpoint:
```bash
(gdb) watch sensor_a_data_0
Hardware watchpoint 1: sensor_a_data_0_data_0
```

If two drivers has the same name for data I do not know how to differentiate.

jeppenodgaard commented on dts/bindings/sensor/nxp,pmc-tmpsns.yaml:
```diff
@@ -0,0 +1,15 @@
+    required: true
+    description: |
+      This should point to an ADC channel (e.g., <&adc0 0>)
+      to  read from the PMC internal temperature sensor.
```
```suggestion
      to read from the PMC internal temperature sensor.
```

ZhaoxiangJin replied:
updated

jeppenodgaard commented on boards/nxp/mimxrt700_evk/board.c:
```diff
@@ -8,6 +8,9 @@
 #include "fsl_clock.h"
 #include <soc.h>
 #include <fsl_glikey.h>
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pmc_tmpsns), okay)
```
```suggestion
#if DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(pmc_tmpsns))
```

jeppenodgaard commented on boards/nxp/mimxrt700_evk/board.c:
```diff
@@ -544,6 +547,12 @@ void board_early_init_hook(void)
 	RESET_ClearPeripheralReset(kACMP0_RST_SHIFT_RSTn);
 #endif
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(pmc_tmpsns), okay)
```
```suggestion
#if DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(pmc_tmpsns))
```

ZhaoxiangJin commented:
@jeppenodgaard, please review

jeppenodgaard commented:
> @jeppenodgaard, please review

LGTM. Other requests needs to be addressed before I can approve.

MaureenHelm reviewed (commented):
Looks ok, but need to resolve the twister failures

ZhaoxiangJin commented:
I can reproduce this issue on zephyr latest main branch, hello @hakehuang, did you catch this issue? Thanks.

ZhaoxiangJin commented:
> Looks ok, but need to resolve the twister failures

I can reproduce this issue on zephyr latest main branch, hello @hakehuang, did you catch this issue? Thanks.

hakehuang commented:
> I can reproduce this issue on zephyr latest main branch, hello @hakehuang, did you catch this issue? Thanks.

Thanks https://github.com/zephyrproject-rtos/zephyr/issues/97020 @ZhaoxiangJin

ZhaoxiangJin commented:
rebased to main, ci issue should be fixed in main.

ZhaoxiangJin commented:
Hello @hakehuang, the CI show new failures, could you please help check again? Thanks!

hakehuang commented on drivers/sensor/nxp/nxp_pmc_tmpsns/nxp_pmc_tmpsns.c:
```diff
@@ -0,0 +1,141 @@
+
+	cm_vref = cm_ctat + (953.36f + calibration) * cm_temp / 2048;
+
+	data->pmc_tmpsns_value = 370.98f * (cm_temp / cm_vref) - 273.15f;
```
Advantages:

Faster execution on processors without FPU
Deterministic behavior (no floating-point rounding issues)
Lower power consumption on embedded systems
Maintains accuracy through careful scaling

code sample with 0.001 precision.

```
// Fixed-point arithmetic configuration
#define TEMP_SCALE_FACTOR         1000    // Scale factor for 3 decimal places
#define TEMP_SCALE_FACTOR_LARGE   1000000 // For intermediate calculations

// Temperature conversion constants (scaled by 1000)
#define TEMP_KELVIN_TO_CELSIUS_SCALED    273150  // 273.15 * 1000
#define TEMP_COEFFICIENT_SCALED          370980  // 370.98 * 1000
#define VREF_BASE_OFFSET_SCALED          953360  // 953.36 * 1000
#define VREF_SCALE_FACTOR                 2048   // Already an integer
#define CALIBRATION_MASK                  0xFF

/**
 * Calculate weighted average for CTAT using integer arithmetic
 * Result is scaled by 1000 for precision
 */
static int32_t calculate_cm_ctat_int(const int16_t *values)
{
    int32_t sum = (2 * values[1]  - values[2] +
                   2 * values[13] - values[12] +
                   2 * values[6]  - values[5] +
                   2 * values[8]  - values[9]);
    
    // Multiply by 1000 for scaling, then divide by 4
    // Equivalent to: sum * 250
    return (sum * 250);
}

/**
 * Calculate weighted average for temperature using integer arithmetic
 * Result is scaled by 1000 for precision
 */
static int32_t calculate_cm_temp_int(const int16_t *values)
{
    int32_t sum = (2 * values[0]  - values[3] +
                   2 * values[14] - values[11] +
                   4 * values[7]  - values[4] - values[10]);
    
    // Multiply by 1000 for scaling, then divide by 4
    // Equivalent to: sum * 250
    return (sum * 250);
}


int32_t get_temperature_millidegrees(sensor_data_t *data, const int16_t *pmc_tmpsns_value)
{
    if (!data || !pmc_tmpsns_value) {
        return INT32_MIN; // Error value
    }

    // Calculate temperature sensor components (scaled by 1000)
    int32_t cm_ctat_scaled = calculate_cm_ctat_int(pmc_tmpsns_value);
    int32_t cm_temp_scaled = calculate_cm_temp_int(pmc_tmpsns_value);
    
    // Extract calibration value
    int32_t calibration = (int8_t)(data->pmc_tmpsns_calibration & CALIBRATION_MASK);
    
    // Calculate reference voltage with calibration
    int64_t vref_numerator = (int64_t)(VREF_BASE_OFFSET_SCALED + calibration * TEMP_SCALE_FACTOR) * cm_temp_scaled;
    int32_t cm_vref_scaled = cm_ctat_scaled + (int32_t)(vref_numerator / (VREF_SCALE_FACTOR * TEMP_SCALE_FACTOR));
    
    if (cm_vref_scaled == 0) {
        return INT32_MIN; // Error value
    }
    
    // Calculate temperature in millidegrees Celsius
    int64_t temp_ratio = ((int64_t)TEMP_COEFFICIENT_SCALED * cm_temp_scaled) / cm_vref_scaled;
    return (int32_t)temp_ratio - TEMP_KELVIN_TO_CELSIUS_SCALED;
}
```

ZhaoxiangJin replied:
Hello @hakehuang, the pmc-tmpsns driver only be used on RTxxx, both of them have FPU. I'll share two sets of data:
the first set was generated using the approach in the PR (without scaling), the second set was derived using the method you suggested. These data were measured using the MIMXRT700-EVK board.

`First set data active mode power consumption (0.43970uA):`
<img width="1877" height="878" alt="image" src="https://github.com/user-attachments/assets/70433add-cc9b-4bbd-9a85-18bd7be24d10" />
`First set data Code execution time (352ns):`
<img width="3505" height="2629" alt="image" src="https://github.com/user-attachments/assets/066f1c8f-7cd5-44d3-ba3f-4e75354a3e08" />


`Second set data active mode power consumption (0.43943uA):`
<img width="1876" height="830" alt="image" src="https://github.com/user-attachments/assets/35b0e656-0df3-45ba-8582-3861e24d1477" />

`Second set data Code execution time (1300ns):`
<img width="3505" height="2629" alt="image" src="https://github.com/user-attachments/assets/0a0dc533-68f7-48e5-a860-9595a1c7b835" />


The data indicates that the power consumption difference between the two approaches is negligible (`0.43970 μA - 0.43943 μA = 0.00027 μA`), while the code execution efficiency varies significantly (`1300 ns - 352 ns = 948 ns`). In real user case, running an additional 948ns in active mode may result in significantly increased power consumption.

But yes, as SoC vendor, we need to give users more options, so I've integrated the method you provided into the driver.

ZhaoxiangJin committed:
drivers: sensor: pmc_tmpsns: Reduce floating-point operations

Reduce floating-point operations in the nxp_pmc_tmpsns.
Provide new calculation method without floating-point operations.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented on samples/sensor/die_temp_polling/boards/mimxrt700_evk_mimxrt798s_cm33_cpu0.overlay:
```diff
@@ -0,0 +1,28 @@
+		zephyr,input-negative = <MCUX_LPADC_CH6B>;
+		zephyr,differential;
+	};
+};
```
updated

ZhaoxiangJin committed:
samples: enable rt700 die_temp_polling sample

1. The updated regex better matches device names
for various temperature sensors. If a device name
contains hyphens, the new regex can match it.

2. enable rt700 die_temp_polling sample

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

MaureenHelm commented on drivers/sensor/nxp/nxp_pmc_tmpsns/Kconfig:
```diff
@@ -0,0 +1,21 @@
+	help
+	  TSENS_CAL is an 8-bit signed calibration constant
+	  retrieved from non-volatile memory. We need this
+	  index to read the fuse to get TSENS_CAL.
```
In the future, you might want to look into using the new [NVMEM subsystem](https://docs.zephyrproject.org/latest/services/nvmem/index.html) for this.

ZhaoxiangJin replied:
Thanks for the reminder. Yes, it's best to use the new subsystem to read FUSE information. I'll first enable that driver and then update the pmc_tmpsns driver.

ZhaoxiangJin replied:
Hello @MaureenHelm, currently, nvmem only supports the EEPROM backend and does not yet support FUSE/OTP backends. Should we block this PR solely because of this limitation? I have already communicated with the nvmem maintainer, and I will integrate the FUSE/OTP backend into nvmem in the future (A significant portion of the work has already been completed, but further review by additional personnel is required.). If everything goes smoothly, I will update the code here at that time.

ZhaoxiangJin replied:
Hello @MaureenHelm, after further consideration, it may take some time to integrate the FUSE backend in NVMEM. Can we not block this PR at this time? Thanks.

MaureenHelm replied:
Meant this as a suggestion for future improvement, not a blocking comment. Sorry for the confusion

ZhaoxiangJin commented on samples/sensor/die_temp_polling/sample.yaml:
```diff
@@ -13,4 +13,4 @@ tests:
       type: one_line
       regex:
-        - "CPU Die temperature\\[[A-Za-z0-9_@]+\\]: [1-9][0-9].[0-9] °C"
+        - "CPU Die temperature\\[[\\-A-Za-z0-9_@]+\\]: [1-9][0-9].[0-9] °C"
```
sorry for missing this, have updated.

ZhaoxiangJin commented:
Hi @MaureenHelm, may i request you review again? Thanks.
